### PR TITLE
ao_avfoundation: added ao driver with AirPlay support

### PR DESCRIPTION
**Pull request by gazsiazasz**
Requires tvOS 11.0, watchOS 4.0, iOS 11.0, macOS 10.13 It has also spatial audio on supported models

Based on @tmm1's work

### CODE REVIEW

**Comment by tmm1** [comment]
I believe @rcombs also revived a version of this driver.

**Review by rcombs — CHANGES_REQUESTED** [review]
I have a similar branch at https://github.com/rcombs/mpv/tree/avfoundation that's still experimental (due to some startup issues that your branch might have fixed; I'll take a look later). Mine also has some additional functionality for eac3 atmos, along with better OS version targeting handling. Hopefully we can get the differences reconciled.

File: audio/out/ao_avfoundation.m [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+
+static int init(struct ao *ao)
+{
+    ao->device_buffer = af_fmt_to_bytes(ao->format) * ao->channels.num * (ao->samplerate / 5);
+
+    struct priv *p = ao->priv;
+
+    if (@available(tvOS 11.0, watchOS 4.0, iOS 11.0, macOS 10.13, *)) {
```
[the remark this answers is not in the document]
Comment (gazsiazasz) [reply audio/out/ao_avfoundation.m]: Increased availability version numbers and added checks.
I don't have currently any working S/PDIF capable Mac at hand so I didn't dig too much into that.

**Comment by MaddTheSane** [comment]
Any update on this?

**Review by rcombs — COMMENTED** [review]
Just realized I had some comments laying around that I'd forgotten to send.

**Comment by gazsiazasz** [comment]
"You don't need S/PDIF or HDMI output; if you pass E-AC-3 with Atmos to AVFoundation, it'll get spatialized on AirPods with head tracking."

It gets spatialized without any tricks if the input is multi-channel.

File: audio/out/ao.c [hunk 2]
```diff
@@ -65,6 +66,9 @@ static const struct ao_driver * const audio_out_drivers[] = {
 #if HAVE_COREAUDIO
     &audio_out_coreaudio,
 #endif
+#if HAVE_AVFOUNDATION
+        &audio_out_avfoundation,
```
[the remark this answers is not in the document]
Comment (gazsiazasz) [reply audio/out/ao.c]: corrected the indentation

File: audio/out/ao_avfoundation.m (hunk 1 above)
[the remark this answers is not in the document]
Comment (gazsiazasz) [reply audio/out/ao_avfoundation.m]: It gets spatialized without any tricks if the input is multi-channel.

**Comment by gazsiazasz** [comment]
@tmm1 please assign another reviewer, @rcombs seems dead :(

**Comment by rcombs** [comment]
Not dead, just laid off from the job I'd been working on this area at.

**Comment by Akemi** [comment]
closing this in favour of #11955.

**Pull request closed without merging**